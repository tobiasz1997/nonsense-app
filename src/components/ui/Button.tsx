import { ButtonHTMLAttributes, cloneElement, FC, ReactElement } from 'react';
import cx from 'classnames';

type Props = {
	size?: 'default' | 'small';
	variant?: 'primary' | 'secondary';
	icon?: ReactElement;
};
const Button: FC<ButtonHTMLAttributes<HTMLButtonElement> & Props> = (props) => {
	const sizes: Record<string, string> = {
		small: 'h-9 text-sm rounded',
		default: 'h-12 rounded'
	};

	const variants: Record<string, string> = {
		primary:
			'text-yellow bg-green-dark hover:bg-green-dark/[0.7] hover:shadow-xl focus:ring-green font-bold px-3',
		secondary:
			'text-green-dark bg-pistachio hover:bg-pistachio/[0.7] hover:shadow-xl focus:ring-yellow font-bold px-3'
	};

	return (
		<button
			{...props}
			className={cx(
				'relative flex w-full items-center justify-center whitespace-nowrap transition focus:outline-none focus:ring-4 disabled:opacity-50',
				props.className,
				sizes[props.size ?? 'default'],
				variants[props.variant ?? 'primary']
			)}
			type={props.type}
			disabled={props.disabled}
		>
			{
				<>
					<span className={'flex h-full w-full items-center justify-center'}>
						{props.icon ? (
							<>
								<span className="flex h-12 w-12 shrink-0 items-center justify-center fill-current px-3">
									{cloneElement(
										props.icon,
										{
											className: 'w-6 h-6'
										},
										null
									)}
								</span>
								<span className="-ml-12 flex w-full justify-center px-3">
									{props.children}
								</span>
							</>
						) : (
							<span>{props.children}</span>
						)}
					</span>
				</>
			}
		</button>
	);
};

export default Button;
