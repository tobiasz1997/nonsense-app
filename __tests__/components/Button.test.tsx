import { render, screen } from '@testing-library/react';
import Button from '@components/ui/Button';

describe('Button', () => {
	const content = '@#$%^&*()qwerty';

	it('should render without crashing', () => {
		const { container } = render(<Button>{content}</Button>);
		expect(container).toMatchSnapshot();
	});

	it('should label be correct', () => {
		render(<Button>{content}</Button>);

		const text = screen.getByText(content);

		expect(text).toBeInTheDocument();
	});
});
