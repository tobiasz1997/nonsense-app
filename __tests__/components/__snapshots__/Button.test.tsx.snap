// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render without crashing 1`] = `
<div>
  <button
    class="relative flex w-full items-center justify-center whitespace-nowrap transition focus:outline-none focus:ring-4 disabled:opacity-50 h-12 rounded text-yellow bg-green-dark hover:bg-green-dark/[0.7] hover:shadow-xl focus:ring-green font-bold px-3"
  >
    <span
      class="flex h-full w-full items-center justify-center"
    >
      <span>
        @#$%^&*()qwerty
      </span>
    </span>
  </button>
</div>
`;
